src/torchjd/sparse/_structured_sparse_tensor.py

```diff
@@ -0,0 +1,278 @@
+        :param strides: Integer (int64) tensor of shape [virtual_ndim, physical_ndim], representing
+            the linear transformation between an index in the physical tensor and the corresponding
+            index in the virtual tensor, i.e. v_index = strides @ p_index.
+        """
+
+        if any(s == 1 for s in physical.shape):
+            raise ValueError(
+                "physical must not contain any dimension of size 1. Found physical.shape="
+                f"{physical.shape}."
+            )
+        if strides.dtype is not torch.int64:
+            raise ValueError(
+                f"strides should be of int64 dtype. Found strides.dtype={strides.dtype}."
+            )
+        if not (strides >= 0).all():
+            raise ValueError(f"All strides must be non-negative. Found strides={strides}.")
+        if strides.shape[1] != physical.ndim:
+            raise ValueError(
+                f"strides should have 1 column per physical dimension. Found strides={strides} and "
+                f"physical.shape={physical.shape}."
+            )
+        if (strides.sum(dim=0) == 0).any():
+            raise ValueError(
+                f"strides should not have any column full of zeros. Found strides={strides}."
+            )
+        groups = get_groupings(list(physical.shape), strides)
+        if any(len(group) != 1 for group in groups):
+            raise ValueError(
+                f"Dimensions must be maximally grouped. Found strides={strides} and "
+                f"groups={groups}"
+            )
+
+        self.physical = physical
+        self.strides = strides
+
+    def to_dense(
+        self, dtype: torch.dtype | None = None, *, masked_grad: bool | None = None
+    ) -> Tensor:
+        assert dtype is None  # We may add support for this later
+        assert masked_grad is None  # We may add support for this later
+
+        if self.physical.ndim == 0:
+            return self.physical
+
+        p_index_ranges = [arange(s) for s in self.physical.shape]
+        p_indices_grid = stack(meshgrid(*p_index_ranges, indexing="ij"))
+
+        # addmm_cuda not implemented for Long tensors => gotta have these tensors on cpu
+        v_indices_grid = tensordot(self.strides, p_indices_grid, dims=1)
+        res = zeros(self.shape, device=self.device, dtype=self.dtype)
+        res[tuple(v_indices_grid)] = self.physical
+        return res
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        kwargs = {} if kwargs is None else kwargs
+
+        if func in cls._HANDLED_FUNCTIONS:
+            return cls._HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+        print_fallback(func, args, kwargs)
+        unwrapped_args = tree_map(unwrap_to_dense, args)
+        unwrapped_kwargs = tree_map(unwrap_to_dense, kwargs)
+        return func(*unwrapped_args, **unwrapped_kwargs)
+
+    def __repr__(self, *, tensor_contents=None) -> str:
```

Comment:
Mypy is unhappy about this typing. I think this is because the return type of `__repr__` in torch is not specified, it expects `Any`. We may want to open a PR (but I could not find the place where it is defined in torch).